src/ledger/mod.rs

```diff
@@ -105,6 +106,77 @@ impl Ledger {
         Ok(())
     }
 
+    #[instrument(name = "ledger.utxo_dropped", skip(self, tx))]
+    pub async fn utxo_dropped(
+        &self,
+        tx: Transaction<'_, Postgres>,
+        tx_id: LedgerTransactionId,
+        detected_txn_id: LedgerTransactionId,
+    ) -> Result<(), LedgerError> {
+        let txs = self
+            .inner
+            .transactions()
+            .list_by_ids(std::iter::once(detected_txn_id))
+            .await?;
+
+        if let Some(UtxoDetectedMeta {
+            account_id,
+            wallet_id,
+            keychain_id,
+            outpoint,
+            satoshis,
+            address,
+            encumbered_spending_fees,
+            confirmation_time,
+        }) = txs[0].metadata()?
+        {
+            let entries = self
+                .inner
+                .entries()
+                .list_by_transaction_ids(vec![detected_txn_id])
+                .await?;
+
+            let mut onchain_incoming_account_id = None;
+            let mut effective_incoming_account_id = None;
+            let mut onchain_fee_account_id = None;
+
+            for entry in entries.into_values().flatten() {
+                match entry.entry_type.as_str() {
+                    "UTXO_DETECTED_UTX_IN_PEN_CR" => {
+                        onchain_incoming_account_id = Some(entry.account_id)
+                    }
+                    "UTXO_DETECTED_LOG_IN_PEN_CR" => {
+                        effective_incoming_account_id = Some(entry.account_id)
+                    }
+                    "UTXO_DETECTED_FR_ENC_DR" => onchain_fee_account_id = Some(entry.account_id),
+                    _ => {}
+                }
+            }
+
+            let params = UtxoDroppedParams {
+                journal_id: txs[0].journal_id,
+                onchain_incoming_account_id: onchain_incoming_account_id.unwrap(),
```

Comment:
You must never - ever -ever ever ever evert ever ever!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Call `unwrap` in production code!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Only in very rare instances is `expect` allowed but even that should be avoided unless you can show clearly in the code above that it is truly an impossible state. Here you are calling a database - that is obviously not a source with a guaranteed outcome.